blpapi.ts

```diff
@@ -0,0 +1,728 @@
+            this.unlisten(messageType, correlatorId);
+        }
+    }
+
+    private sessionTerminatedHandler(ev: any): void {
+        log('Session terminating');
+        trace(ev);
+
+        _([{prop: 'eventListeners', cleanupFn: (eventName: string): void => {
+            this.removeAllListeners(eventName);
+         }},
+         {prop: 'requests', cleanupFn: (k: string): void => {
+            this.requests[k](new Error('session terminated'));
+         }},
+         {prop: 'subscriptions', cleanupFn: (k: string): void => {
+            this.subscriptions[k].emit('error', new Error('session terminated'));
+         }}
+       ]).forEach((table: {prop: string; cleanupFn: (s: string) => void}): void => {
+            Object.getOwnPropertyNames(this[table.prop]).forEach((key: string): void => {
+                table.cleanupFn(key);
+            });
+            this[table.prop] = null;
+        });
+
+        if (!this.stopped) {
+            this.stopped = Promise.resolve();
+        }
+
+        // tear down the session
+        this.invoke(this.session.destroy);
+        this.session = null;
+
+        // emit event to any listeners
+        this.emit('SessionTerminated', ev.data);
+        log('Session terminated');
+    }
+
+    private doRequest(uri: string,
+                      requestName: string,
+                      request: any,
+                      execute: (correlatorId: number) => void,
+                      callback: IRequestCallback): void
+    {
+        // TODO: Review all calls to 'validateSession' which throws bypassing
+        // the expected error reporting mechanism.
+        try {
+            this.validateSession();
+        } catch (ex) {
+            callback(ex);
+            return;
+        }
+
+        var correlatorId = this.nextCorrelatorId();
+        this.requests[correlatorId] = callback;
+
+        this.openService(uri).then((): void => {
+            log(util.format('Request: %s|%d', requestName, correlatorId));
+            trace(request);
+            execute(correlatorId);
+            assert(requestName in REQUEST_TO_RESPONSE_MAP,
+                   util.format('Request, %s, not handled', requestName));
+            this.listen(REQUEST_TO_RESPONSE_MAP[requestName],
+                        correlatorId,
+                        this.requestHandler.bind(this, callback));
+        }).catch((ex: Error): void => {
+            delete this.requests[correlatorId];
+            callback(ex);
+        });
+    }
+
+    // PRIVATE ACCESSORS
+    private validateSession(): void {
+        if (this.stopped) {
+            throw new Error('session terminated');
+        }
+    }
+
+    // CREATORS
+    constructor(opts: ISessionOpts) {
+        super();
+
+        this.session = new blpapijs.Session(opts);
+        this.once('SessionTerminated', this.sessionTerminatedHandler.bind(this));
+        trace(opts);
+        var that = this;
+        this.session.emit = (): boolean => {
```

Comment:
Right! We should avoid surfacing to the user the events generated at the binding layer level. But because the binding layer Session type doesn't provide a way to listen to the event it generates we need a thin adaptor type that behaves as a 'regular' event emitter/listener.